### PR TITLE
Add authn for graphql and http admin endpoints

### LINKED ISSUE #4758: Export endpoint doesn't require authentication
### What version of Dgraph are you using?
v2.0.0-beta1-26-g2ad482aff

### Have you tried reproducing the issue with the latest release?
Yes

### Steps to reproduce the issue (command/config used to run Dgraph).
Just send request to export endopoint, it will start export.

### Expected behaviour and actual result.
It should require authentication.

### PR DESCRIPTION
**Pull request by abhimanyusinghgaur**
### Description.

This PR adds authentication to following endpoints:
* /admin/backup (http & graphql)
* /admin/config/lru_mb (http [GET & PUT] & graphql [query & mutation])
* /admin/draining (http & graphql)
* /admin/export (http & graphql)
* /admin/shutdown (http & graphql)
* /admin/restore (graphql only)
* /admin/listBackups (graphql only)

Now, all the above http endpoints and their corresponding graphql versions have following kinds of auth:
* IP White-listing, if --whitelist flag is passed to alpha
* Poor-man's auth, if `--auth_token` flag is passed to alpha
* Guardian only access, if ACL is enabled

This PR also adds query for `config` in graphql admin, as it was missing earlier.

In addition to above points:
* All the `/admin` endpoints apply Poor-man's auth check at http level itself, while other auth checks are routed through graphql resolvers.
* GraphQL Resolvers for health/state and the ones related to ACL User/Group have IP whitelisting middleware applied, while dgraph handles Guardian auth for them.
* `/alter` has the existing behaviour of checking only Poor-man's and Guardian auth.
* GraphQL Resolvers related to schema don't apply IP whitelisting as to keep them in sync with `/alter`. They do apply Guardian auth.
* Any GraphQL admin introspection queries don't require IP whitelisting or Guardian auth.

### GitHub Issue or Jira number.

Fixes #4758.

### Other components or 3rd party tools affected (or regression areas).

No

### Affected releases.

\> 20.03.0

### Changelog tags.

added authentication for admin endpoints

### Please indicate if this is a breaking change.

No

### Please indicate if this is an enterprise feature.

No

### Please indicate if documentation needs to be updated.

Yes

### Please indicate if end to end testing is needed.

Yes, for the whitelisting and poor-man's ACL part.

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/5162)

 
 
 
 
 

Docs Preview: [Dgraph Preview](https://dgraph-8fe2324532-64985.surge.sh)

### CODE REVIEW

File: dgraph/cmd/alpha/admin.go [hunk 1]
```diff
@@ -25,54 +25,83 @@ import (
...
+		x.SetStatus(w, x.ErrorUnauthorized, "Invalid X-Dgraph-AuthToken")
+		return false
+	}
+	return true
+}
+
 // handlerInit does some standard checks. Returns false if something is wrong.
-func handlerInit(w http.ResponseWriter, r *http.Request, allowedMethods map[string]bool) bool {
+func handlerInit(w http.ResponseWriter, r *http.Request, allowedMethods map[string]bool,
+	allowOnlyGuardians bool) bool {
```
Comment (gja) [inline dgraph/cmd/alpha/admin.go:55]: should we be clubbing allowedMethods and allowOnlyGuardians into a type?

Comment (gja) [reply dgraph/cmd/alpha/admin.go:55]: Also, can we rename handlerInit to something like checkAuth?

File: dgraph/cmd/alpha/admin.go [hunk 2]
```diff
@@ -25,54 +25,83 @@ import (
...
 	if _, ok := allowedMethods[r.Method]; !ok {
```
Comment (gja) [inline dgraph/cmd/alpha/admin.go:56]: can't we move these allowedMethods type checks into some sort of router or something? I know we do use httprouter, why aren't 405s moved over there?

File: dgraph/cmd/alpha/admin.go [hunk 3]
```diff
@@ -25,54 +25,83 @@ import (
...
 		x.SetStatus(w, x.ErrorInvalidMethod, "Invalid method")
+		w.WriteHeader(http.StatusMethodNotAllowed)
```
Comment (gja) [inline dgraph/cmd/alpha/admin.go:58]: Isn't this a bit strange? Middlewares should either do the `return true/false` pattern, or they should `call next() on success`. This feels like it's mixing both

File: dgraph/cmd/alpha/run.go [hunk 4]
```diff
@@ -453,10 +453,24 @@ func setupServer(closer *y.Closer) {
 	// TODO: Figure out what this is for?
 	http.HandleFunc("/debug/store", storeStatsHandler)
 
-	http.HandleFunc("/admin/shutdown", shutDownHandler)
-	http.HandleFunc("/admin/draining", drainingHandler)
-	http.HandleFunc("/admin/export", exportHandler)
-	http.HandleFunc("/admin/config/lru_mb", memoryLimitHandler)
+	http.Handle("/admin/shutdown", adminAuthHandler(http.HandlerFunc(shutDownHandler),
+		adminAuthOptions{allowedMethods: map[string]bool{
```
Comment (abhimanyusinghgaur) [inline dgraph/cmd/alpha/run.go:457]: change the type for allowedMethods to a alias type

File: graphql/admin/admin.go [hunk 5]
```diff
@@ -408,22 +408,24 @@ func newAdminResolver(
...
 
-	adminMutationResolvers := map[string]resolve.MutationResolverFunc{
-		"backup":   resolveBackup,
-		"config":   resolveConfig,
-		"draining": resolveDraining,
-		"export":   resolveExport,
-		"login":    resolveLogin,
-		"restore":  resolveRestore,
-		"shutdown": resolveShutdown,
+	adminMutationResolvers := map[string]resolve.MutationResolver{
```
Comment (gja) [inline graphql/admin/admin.go:411]: Please make this configuration into a map (or even a configuration file).

"backup": applyMiddleWare(resolveBackup, middleware["backups"])

Comment (gja) [reply graphql/admin/admin.go:411]:
```go
middleware := {
	"backup": &MiddlewareOptions{
		poorman: true,
		ipwhitelist: true
	}
}

http.handle("/backup", AdminAuthHandler(handleBackup, middleware["backups"]))
foo.AddMutationResolver(AdminAuthResolver(someResolver(qtx, etc...), middleware["backups"]))
```

File: ee/acl/acl_test.go [hunk 6]
```diff
@@ -1821,6 +1801,219 @@ func TestHealthForAcl(t *testing.T) {
 	}
 }
 
+func TestAclForAdminEndpoints(t *testing.T) {
+	tcases := []struct {
```
Comment (gja) [inline ee/acl/acl_test.go:1805]: move the constant out of the function

**Pull request merged**